### ISSUE TITLE
nunit.engine can find incorrect nunit.framework assembly in some situations

### ISSUE THREAD

**Issue by cyanite** [nunit-legacy/nunitv2#74]
In some situations, involving an app domain with both nunit 3 and 2, nunit.engine (which is part of nunit 2) will load the wrong version of nunit.framework, or rather, will find the wrong version of the assembly, which has already been loaded.

This can be reproduced using the nunit3-console runner as follows:

1. An NUnit project file with two assemblies. The first, A.dll, should reference NUnit 3 framework and the second, B.dll, should reference NUnit 2 framework. Note: In order to make this work, copy the two nunit.framework dll's to new file names, say n2.dll and n3.dll, and write an app.config file like this:
```
<configuration>
  <runtime>
    <assemblyBinding xmlns="urn:schemas-microsoft-com:asm.v1">
      <dependentAssembly>
        <assemblyIdentity name="nunit.framework" publicKeyToken="2638cd05610744eb" />
        <codeBase version="3.10.1.0" href="n3.dll" />
      </dependentAssembly>
      <dependentAssembly>
        <assemblyIdentity name="nunit.framework" publicKeyToken="96d09a1eb7f44a77" />
        <codeBase version="2.6.4.14350" href="n2.dll" />
      </dependentAssembly>
    </assemblyBinding>
  </runtime>
</configuration>
```
2. The NUnit project file should use a single process and app domain and refer to the app.config file created above (as well as the two test assemblies, of course), e.g.
```
<NUnitProject>
  <Settings activeconfig="Default" processModel="Separate" domainUsage="Single" />
    <Config configfile="app.config" binpathtype="Auto" name="Default">
      <assembly path="A.dll" />
      <assembly path="B.dll" />
  </Config>
</NUnitProject>
```
3. Both assemblies should declare a custom test action, and decorate at least one test with it.
5. The NUnit v2 driver should be available.

Now, when running nunit3-console on the project file, test actions for assembly B will not be executed. If you swap the assembly order, the will execute correctly.

This happens because nunit.engine just looks for the first loaded assembly called "nunit.framework", but in the above situation this will be v3. I created a pull request to only look for nunit.framework v2, which fixes the issue.

NUnit 3 handles this situation correctly already.

**Comment by CharliePoole** [comment]
OK... I see why it happens now. It's frankly a bit pathological and I would probably have rejected it as an issue if it didn't come with a PR. I'd just say "Well don't do that!" :smiley: 

What appears to have happened here is that the user insisted on running two tests with different requirements in the same process and appdomain. Then to make that work, binding redirects were added. I'd say at that point one could ask "Why am I doing this?" There's usually no good reason to force two test assemblies to run together. In fact, one usually wants to have them isolated.

All that said, since you have the code to fix it already, I'll give it a review. Since NUnit __allows__ you to run the tests in the same AppDomain, it should not make the assumption (like the current code does) that only one copy of the framework has been loaded.

I'm travelling so my time on this app is a bit spotty, but I should be able to get back to it in a day or two.

**Comment by cyanite** [comment]
> What appears to have happened here is that the user insisted on running two tests with different requirements in the same process and appdomain. Then to make that work, binding redirects were added. I'd say at that point one could ask "Why am I doing this?" There's usually no good reason to force two test assemblies to run together. In fact, one usually wants to have them isolated.

Right. Our motivation is that we have a very large monolith which is divided into like 700 assemblies with a lot of tests, many of the integration tests. We're upgrading gradually to NUnit 3, but it's not always easy due to very non-trivial use of non-static test case sources.

Because the integration tests can involve a fairly large startup cost and because it consumes a lot of memory, we try to run all tests in a single domain. We then use app.config to redirect identically named assemblies as above, and run a mixed NUnit 2/3 suite (if you perhaps recall, I have made earlier (accepted) pull requests to both NUnit 3 and 2 to load assemblies correctly in this situation.)

The present problem happened by accident once we converted the top assembly in the NUnit project.

**Comment by CharliePoole** [comment]
Sorry, these days I often model a stateless protocol in discussing issues. :smile: 

Your case is, in fact, the very rare case for which we continue to support forcing things into the same AppDomain in NUnit 3. Personally, I would remove that feature otherwise... or if we could provide some equivalent functionality for coordinating across AppDomains and Processes.

**Comment by cyanite** [comment]
> Your case is, in fact, the very rare case for which we continue to support forcing things into the same AppDomain in NUnit 3.

Sure, it's pretty unusual, but I think there are other less extreme usecases of single domain.

And, with these fixes (including my previous), it works well, so why remove it. Also, the assembly loading was technically broken before (what with the partial load and all), even if it would never actually show in "normal" use ;).

**Comment by CharliePoole** [comment]
I'll close but I do have one more question about what you are doing... 

You are running NUnit V2 and NUnit 3 tests together using nunit3-console, and there fore the NUnit engine. The NUnit engine will load the NUnit V2 Framework driver for the V2 assembly.

We just fixed the NUnit V2 legacy project, which (if the above is correct) you are not using. How will this help you? In order for your situation to work, it seems that the NUnit team has to fix the NUnit V2 Framework Driver as well in precisely the same way. AFAIK, the V2 framework driver is still using NUnit 2.6.4 core assembly.

**Comment by CharliePoole** [comment]
Correction: the V2 framework driver is using the 2.7.0 package that I create. I either never knew that or forgot it. :smile: So, what has to happen is...

1. I create a 2.7.1 package
2. They do a new extension package that references my 2.7.1
3. Users install that new extension separately, until
4. The NUnit team releases another new console/engine package that includes the updated extension.

Pulling in @rprouse so somebody knows about this.

**Comment by CharliePoole** [comment]
@cyanite FYI, this fix is in the 2.7.1 release I just did.
